packages/hash/shared/src/ProsemirrorManager.ts

```diff
@@ -0,0 +1,530 @@
+         *       information
+         */
+        toDOM(node) {
+          if (node.textContent?.length > 0) {
+            return ["span", { "data-hash-type": "component" }, 0];
+          } else {
+            return ["span", { "data-hash-type": "component" }];
+          }
+        },
+      });
+    });
+
+    if (this.componentNodeViewFactory && this.view) {
+      this.view.setProps({
+        nodeViews: {
+          // Private API
+          ...(this.view as any).nodeViews,
+          [componentId]: this.componentNodeViewFactory(meta),
+        },
+      });
+    }
+  }
+
+  /**
+   * Defining a new type of block in Prosemirror. Designed to be cached so
+   * doesn't need to request the block multiple times
+   *
+   * @todo support taking a signal
+   */
+  async defineBlockByComponentId(
+    componentId: string,
+    options?: { bustCache: boolean },
+  ): Promise<HashBlock> {
+    const block = await fetchBlock(componentId, {
+      bustCache: !!options?.bustCache,
+    });
+
+    this.defineBlock(block);
+
+    return block;
+  }
+
+  /**
+   * Blocks need to be defined before loading into Prosemirror
+   */
+  async ensureBlocksDefined(componentIds: string[] = []) {
+    return Promise.all(
+      componentIds.map((componentId) =>
+        this.defineBlockByComponentId(componentId),
+      ),
+    );
+  }
+
+  /**
+   * @note targetComponentId must already be defined
+   * @see defineBlock
+   * @see defineBlockByComponentId
+   */
+  renderBlock(
+    targetComponentId: string,
+    entityStore: EntityStore | null = null,
+    draftBlockId?: string | null,
+  ) {
+    this.assertBlockDefined(targetComponentId);
+
+    const blockEntity =
+      draftBlockId &&
+      isBlockCompatible(targetComponentId, draftBlockId, entityStore)
+        ? entityStore?.draft[draftBlockId]
+        : null;
+
+    const childDraftId = isDraftBlockEntity(blockEntity)
+      ? blockEntity.properties.entity.draftId
+      : null;
+
+    const blockData =
+      draftBlockId && entityStore
+        ? getBlockChildEntity(draftBlockId, entityStore)
+        : null;
+
+    const content =
+      blockData && isTextEntity(blockData)
+        ? childrenForTextEntity(blockData, this.schema)
+        : [];
+
+    return this.schema.nodes.block!.create({}, [
+      this.schema.nodes.entity!.create(
+        { draftId: draftBlockId },
+        this.schema.nodes.entity!.create(
+          {
+            draftId: childDraftId,
+          },
+          [this.schema.nodes[targetComponentId]!.create({}, content)],
+        ),
+      ),
+    ]);
+  }
+
+  private assertBlockDefined(componentId: string) {
+    if (!this.schema.nodes[componentId]) {
+      throw new Error("Block must already be defined");
+    }
+  }
+
+  async loadPage(currentState: EditorState<Schema>, entities: BlockEntity[]) {
+    const store = createEntityStore(
+      entities,
+      entityStorePluginState(currentState).store.draft,
+    );
+
+    const newNodes = await Promise.all(
```

Comment:
Perhaps we should just call it `nodes`. `newNodes` makes it feel like it's only new blocks that are defined here, whereas it's actually all blocks in the page

Reply:
[`a59d29a` (#872)](https://github.com/hashintel/hash/pull/872/commits/a59d29af479fb27d66fc480bc6f8fd8ea2cfd771)